### PR TITLE
Reporting: Wait for repeat panel clones to settle before signaling render readiness

### PR DESCRIPTION
**Pull request by grafakus**
## Goal

Fixes https://github.com/grafana/support-escalations/issues/23324 by ensuring that the image renderer captures reports only after repeated panels have finished loading, instead of racing ahead when a repeat variable completes.

See the last section below for an alternative approach/PR.

## Problem

Reports for dashboards using panel/row/tab repeats could be captured before the repeated panels had any data, producing empty or partially rendered PDFs (race described in #128536 + customer escalation).

The root cause is a race in how readiness was signaled:
- the Scenes profiler (`SceneRenderProfiler`) observes the `SceneQueryController`'s running-query count
- once the count reaches 0, it initiates a 2s tail-recording window (`POST_STORM_WINDOW`) and, provided no query registers during that interval, finalizes the profile
- it dispatches the interaction-complete notification to its observers (`dashboard_view` event)
- `ReportRenderReadinessObserver` subscribes to that notification and unconditionally propagates it to the image renderer as `REPORT_RENDER_COMPLETE`

With repeats, this termination occurs prematurely. When a repeat variable query finishes, the running-query count briefly drops to 0 _before_ the repeated panel clones are created and register their own queries. But React still has to mount the (potentially large) list of repeated panel components into the DOM before any of their queries start, so the gap can exceed the Scenes profiler's 2 s post-storm window and the profiler concludes `dashboard_view` prematurely.

## Approach

Instead of forwarding the completion event immediately, the observer now holds a reference to the `DashboardScene` and verifies the scene graph has actually settled before sending `REPORT_RENDER_COMPLETE`:

1. No queries running (`window.__grafanaRunningQueryCount`).
2. Every repeater has finished its work: its variable is done loading, its clones are created, and, for row/tab repeats, React has mounted every clone.
3. Every `VizPanel` that is mounted or about to mount (active, or child of an active repeater, i.e. freshly created panel-repeat clones) has no loading variable dependency and resolved data (`Done` / `Error` / `Streaming`). Panels that will never mount, collapsed rows, inactive tabs are excluded so they cannot stall settlement.

If not settled, it polls every 500 ms, with a 20s cap as a safety valve so a stuck panel degrades to the old behavior (capture whatever has loaded).

## Key changes

1. `ReportRenderReadinessObserver.ts` contains the new settlement logic. Before reviewing, note:
   - Repeaters are detected by the shape of their state, not by importing `DashboardGridItem`, `RowItem`, etc. Runtime imports would create an import cycle with `dashboard-scene`. This covers the default grid, auto grid, and rows/tabs layouts.
   - Row/tab clone panels are invisible to the per-panel check until React mounts them, so the repeater stays pending until all clones are active.
   - `RowRepeaterBehavior` keeps its clone list private. We detect pending clones as inactive `SceneGridRow`s with `repeatSourceKey` under an active layout, then wait for those panels' data like any other panel.
   - Panels are collected with a full scene-graph walk (`sceneGraph.findAllObjects`) because `getVizPanels()` skips repeat clones.
   - A panel with `undefined` data counts as not settled. This is the state of a fresh, unmounted repeat clone, which is exactly the race being fixed.
   - The 20s max wait stays below the renderer's `--browser.readiness.timeout` (30 s by default). If we waited longer, the render would fail with a timeout instead of producing a best-effort report.
   - A loading variable only blocks panels that depend on it (`hasVariableDependencyInLoadingState`). Unrelated loading variables are ignored.
   - Repeaters that will never clone don't block the report: row/tab repeaters with a missing or non-multi-value variable are skipped.
2. New integration test `ReportRenderReadinessObserver.repeats.test.ts` reproduces the race end to end.

## How to test

Setup: a dashboard with a panel repeated by a multi-value variable (ideally a slow variable query and/or slow panel queries to widen the race window), rendering via grafana-image-renderer report/PDF generation.

- Generate a report for the repeat dashboard: all repeated panel instances show data (previously some could be blank).
- Repeat with rows/tabs repeats (auto grid and default grid layouts).
- Dashboard without repeats: report renders as before, no added delay beyond one settlement check.
- Stuck-panel edge case (e.g. a query that never resolves): the ready signal is sent after ~20 s and the report captures whatever has loaded, instead of the renderer timing out.
- Unit tests: `yarn test public/app/features/dashboard/services/ReportRenderReadinessObserver`

## Alternative fix

The root can also be addressed in `@grafana/scenes`. A proper fix (repeaters registering an entry with the query controller until their clones mount) would benefit all profiler consumers and let us simplify the polling added here. 

**See this [alternative draft PR](https://github.com/grafana/grafana/pull/129758).**

### CODE REVIEW

**Comment by grafakus** [comment]
cursor review

File: public/app/features/dashboard/services/ReportRenderReadinessObserver.ts [hunk 1]
```diff
@@ -13,32 +25,272 @@ interface RenderBindingMessage<T extends MessageEventType> {
...
+ * it has before sending `REPORT_RENDER_COMPLETE`.
  */
 export class ReportRenderReadinessObserver implements performanceUtils.ScenePerformanceObserver {
-  onDashboardInteractionComplete = (data: performanceUtils.DashboardInteractionCompleteData): void => {
-    if (data.interactionType === 'dashboard_view') {
+  #scene: DashboardSceneLike | undefined;
+  #pollTimeoutId: ReturnType<typeof setTimeout> | undefined;
+  #deadline = 0;
+
+  public setScene(scene: DashboardSceneLike | undefined): void {
```
Comment (tskarhed) [inline public/app/features/dashboard/services/ReportRenderReadinessObserver.ts:224]: While this might be technically feasible, I think this is the wrong place to implement this functionality. This should be addressed in dashboard rendering, so that when repeating panels are completed, the `dashboard_view` is emitted. So maybe the fix in scenes is more suitable?

Comment (grafakus) [reply public/app/features/dashboard/services/ReportRenderReadinessObserver.ts:224]: Indeed, that's what the alternative PRs (#129758 + Scenes https://github.com/grafana/scenes/pull/1593) do, but only for report rendering, because unconditionally deferring `dashboard_view` until all the repeats settle would change our `dashboard_render` load-time metrics/KPIs.

File: public/app/features/dashboard/services/ReportRenderReadinessObserver.ts [hunk 2]
```diff
@@ -13,32 +25,272 @@ interface RenderBindingMessage<T extends MessageEventType> {
...
+      visit(child);
+    });
+  };
+
+  visit(root);
+  return hit;
+}
+
+/** True when this object sits under a default-grid row-repeat clone whose layout is active. */
+function isUnderActiveDefaultGridRowClone(panel: SceneObject): boolean {
```
Comment (ifrost) [inline public/app/features/dashboard/services/ReportRenderReadinessObserver.ts:78]: Have you tested collapsed rows? I haven't tried but from the code it seems it's not checking if the row is collapsed and that would cause always hitting 20s fallback as collapsed panels are not rendered.

**Review by ifrost — COMMENTED** [review]
Maybe just me, but this feels quite obscure and fragile. Have you thought about any alternative solutions to the root problem?

File: public/app/features/dashboard/services/ReportRenderReadinessObserver.ts (hunk 2 above)
Comment (grafakus) [reply public/app/features/dashboard/services/ReportRenderReadinessObserver.ts:78]: I'm not sure it's an issue since [ReportGridRenderer uncollapses all rows](https://github.com/grafana/grafana-enterprise/blob/main/src/public/reports/dashboard/ReportGridRenderer.tsx#L73). But worth double-checking

**Comment by grafakus** [comment]
> Maybe just me, but this feels quite obscure and fragile. Have you thought about any alternative solutions to the root problem?

Yes, see "Alternative fix" in the description: [#129758](https://github.com/grafana/grafana/pull/129758) + [grafana/scenes#1593](https://github.com/grafana/scenes/pull/1593).

This PR has the advantage of scoping the fix to report rendering only: no impact on `dashboard_view` timing and no scenes release needed. But, as mentioned earlier today, it feels overly complex and fragile.
